File: lab5/lab5_1/Python/lab5_1.py
import tkinter as tk

# ── Minimal Modern Palette ────────────────────────────────────────────────────
BG          = "#101014"
PANEL       = "#17181f"
PANEL_2     = "#1d1f27"
BORDER      = "#2a2d37"
ACCENT      = "#7c8cff"
ACCENT_HOV  = "#93a0ff"
TEXT        = "#f3f4f6"
TEXT_DIM    = "#9aa0ad"
SUCCESS     = "#4ade80"
WARN        = "#facc15"
ERROR       = "#f87171"
ENTRY_BG    = "#111319"


class SegmentedToggle(tk.Frame):
    def __init__(self, parent, callback=None, **kwargs):
        super().__init__(parent, bg=PANEL, **kwargs)
        self.callback = callback
        self.mode = "A"

        self.configure(highlightthickness=1, highlightbackground=BORDER)

        self.btn_a = tk.Label(
            self, text="°C → °F",
            bg=ACCENT, fg="white",
            font=("Segoe UI", 10, "bold"),
            padx=24, pady=10,
            cursor="hand2"
        )
        self.btn_b = tk.Label(
            self, text="°F → °C",
            bg=PANEL, fg=TEXT_DIM,
            font=("Segoe UI", 10, "bold"),
            padx=24, pady=10,
            cursor="hand2"
        )

        self.btn_a.pack(side="left", expand=True, fill="both")
        self.btn_b.pack(side="left", expand=True, fill="both")

        self.btn_a.bind("<Button-1>", lambda e: self.set_mode("A"))
        self.btn_b.bind("<Button-1>", lambda e: self.set_mode("B"))

    def set_mode(self, mode):
        self.mode = mode
        if mode == "A":
            self.btn_a.config(bg=ACCENT, fg="white")
            self.btn_b.config(bg=PANEL, fg=TEXT_DIM)
        else:
            self.btn_a.config(bg=PANEL, fg=TEXT_DIM)
            self.btn_b.config(bg=ACCENT, fg="white")

        if self.callback:
            self.callback(mode)


class ModernButton(tk.Label):
    def __init__(self, parent, text, command=None, primary=True, **kwargs):
        super().__init__(
            parent,
            text=text,
            font=("Segoe UI", 10, "bold"),
            padx=16,
            pady=12,
            cursor="hand2",
            **kwargs
        )
        self.command = command
        self.primary = primary

        if primary:
            self.default_bg = ACCENT
            self.hover_bg = ACCENT_HOV
            self.default_fg = "white"
            self.default_border = ACCENT
        else:
            self.default_bg = PANEL_2
            self.hover_bg = "#252833"
            self.default_fg = TEXT
            self.default_border = BORDER

        self.configure(
            bg=self.default_bg,
            fg=self.default_fg,
            highlightthickness=1,
            highlightbackground=self.default_border
        )

        self.bind("<Enter>", self._on_enter)
        self.bind("<Leave>", self._on_leave)
        self.bind("<Button-1>", self._on_click)

    def _on_enter(self, _):
        self.config(bg=self.hover_bg)

    def _on_leave(self, _):
        self.config(bg=self.default_bg)

    def _on_click(self, _):
        if self.command:
            self.command()


class TemperatureConverter(tk.Tk):
    WIN_W = 420
    WIN_H_NORMAL = 430
    WIN_H_RESULT = 560

    def __init__(self):
        super().__init__()
        self.title("Temperature Converter")
        self.configure(bg=BG)
        self.resizable(False, False)

        self._mode = "A"

        self._build_ui()
        self._center(self.WIN_H_NORMAL)

    def _build_ui(self):
        outer = tk.Frame(self, bg=BG)
        outer.pack(fill="both", expand=True, padx=24, pady=24)

        # Main card
        self.card = tk.Frame(
            outer,
            bg=PANEL,
            highlightthickness=1,
            highlightbackground=BORDER
        )
        self.card.pack(fill="both", expand=True)

        # Header
        tk.Label(
            self.card,
            text="Temperature Converter",
            bg=PANEL,
            fg=TEXT,
            font=("Segoe UI", 20, "bold")
        ).pack(anchor="w", padx=24, pady=(22, 4))

        tk.Label(
            self.card,
            text="Convert values instantly with a clean minimal interface.",
            bg=PANEL,
            fg=TEXT_DIM,
            font=("Segoe UI", 9)
        ).pack(anchor="w", padx=24, pady=(0, 18))

        divider = tk.Frame(self.card, bg=BORDER, height=1)
        divider.pack(fill="x", padx=24, pady=(0, 18))

        # Conversion label
        tk.Label(
            self.card,
            text="Conversion",
            bg=PANEL,
            fg=TEXT_DIM,
            font=("Segoe UI", 9)
        ).pack(anchor="w", padx=24)

        self.toggle = SegmentedToggle(self.card, callback=self._on_mode)
        self.toggle.pack(fill="x", padx=24, pady=(8, 18))

        # Input label
        self.input_label = tk.Label(
            self.card,
            text="Temperature (°C)",
            bg=PANEL,
            fg=TEXT_DIM,
            font=("Segoe UI", 9)
        )
        self.input_label.pack(anchor="w", padx=24)

        # Entry shell
        entry_shell = tk.Frame(
            self.card,
            bg=ENTRY_BG,
            highlightthickness=1,
            highlightbackground=BORDER
        )
        entry_shell.pack(fill="x", padx=24, pady=(8, 8))

        self.var = tk.StringVar()
        self.entry = tk.Entry(
            entry_shell,
            textvariable=self.var,
            bg=ENTRY_BG,
            fg=TEXT,
            insertbackground=TEXT,
            relief="flat",
            bd=0,
            font=("Segoe UI", 16),
            highlightthickness=0
        )
        self.entry.pack(fill="x", padx=14, pady=14)
        self.entry.bind("<Return>", lambda e: self._convert())
        self.entry.bind("<FocusIn>", lambda e: entry_shell.config(highlightbackground=ACCENT))
        self.entry.bind("<FocusOut>", lambda e: entry_shell.config(highlightbackground=BORDER))

        # Error label
        self.err = tk.StringVar()
        self.err_label = tk.Label(
            self.card,
            textvariable=self.err,
            bg=PANEL,
            fg=ERROR,
            font=("Segoe UI", 9)
        )
        self.err_label.pack(anchor="w", padx=24, pady=(0, 12))

        # Convert button
        self.convert_btn = ModernButton(
            self.card,
            text="Convert",
            command=self._convert,
            primary=True
        )
        self.convert_btn.pack(fill="x", padx=24, pady=(0, 18))

        # Result card
        self.result_frame = tk.Frame(
            self.card,
            bg=PANEL_2,
            highlightthickness=1,
            highlightbackground=BORDER
        )

        top_row = tk.Frame(self.result_frame, bg=PANEL_2)
        top_row.pack(fill="x", padx=18, pady=(16, 6))

        tk.Label(
            top_row,
            text="Result",
            bg=PANEL_2,
            fg=TEXT_DIM,
            font=("Segoe UI", 9)
        ).pack(side="left")

        self.status_var = tk.StringVar()
        self.status_lbl = tk.Label(
            top_row,
            textvariable=self.status_var,
            bg=PANEL_2,
            fg=WARN,
            font=("Segoe UI", 9, "bold")
        )
        self.status_lbl.pack(side="right")

        value_row = tk.Frame(self.result_frame, bg=PANEL_2)
        value_row.pack(fill="x", padx=18, pady=(0, 6))

        self.result_val = tk.StringVar()
        self.result_unit = tk.StringVar()

        tk.Label(
            value_row,
            textvariable=self.result_val,
            bg=PANEL_2,
            fg=TEXT,
            font=("Segoe UI", 28, "bold")
        ).pack(side="left")

        tk.Label(
            value_row,
            textvariable=self.result_unit,
            bg=PANEL_2,
            fg=ACCENT,
            font=("Segoe UI", 13, "bold")
        ).pack(side="left", padx=(8, 0), pady=(8, 0))

        self.detail_var = tk.StringVar()
        tk.Label(
            self.result_frame,
            textvariable=self.detail_var,
            bg=PANEL_2,
            fg=TEXT_DIM,
            font=("Segoe UI", 9)
        ).pack(anchor="w", padx=18, pady=(0, 16))

        self.reset_btn = ModernButton(
            self.card,
            text="Reset",
            command=self._reset,
            primary=False
        )

    def _on_mode(self, mode):
        self._mode = mode
        self.err.set("")
        self._hide_result()

        if mode == "A":
            self.input_label.config(text="Temperature (°C)")
        else:
            self.input_label.config(text="Temperature (°F)")

    def _convert(self):
        raw = self.var.get().strip()

        if not raw:
            self.err.set("Enter a temperature value.")
            self._hide_result()
            return

        try:
            value = float(raw)
        except ValueError:
            self.err.set("Please enter a valid number.")
            self._hide_result()
            return

        self.err.set("")

        if self._mode == "A":
            converted = (value * 9 / 5) + 32
            out_unit = "°F"
            detail = f"{value:.2f} °C converted to Fahrenheit"
        else:
            converted = (value - 32) * 5 / 9
            out_unit = "°C"
            detail = f"{value:.2f} °F converted to Celsius"

        if converted > 100:
            status = "Too High"
            status_color = WARN
        else:
            status = "Normal"
            status_color = SUCCESS

        self.result_val.set(f"{converted:.2f}")
        self.result_unit.set(out_unit)
        self.detail_var.set(detail)
        self.status_var.set(status)
        self.status_lbl.config(fg=status_color)

        self._show_result()

    def _show_result(self):
        self._center(self.WIN_H_RESULT)
        self.result_frame.pack(fill="x", padx=24, pady=(0, 12))
        self.reset_btn.pack(fill="x", padx=24, pady=(0, 22))

    def _hide_result(self):
        self._center(self.WIN_H_NORMAL)
        self.result_frame.pack_forget()
        self.reset_btn.pack_forget()

    def _reset(self):
        self.var.set("")
        self.err.set("")
        self.result_val.set("")
        self.result_unit.set("")
        self.detail_var.set("")
        self.status_var.set("")
        self.entry.focus_set()
        self._hide_result()

    def _center(self, height):
        self.geometry(f"{self.WIN_W}x{height}")
        self.update_idletasks()
        x = (self.winfo_screenwidth() - self.WIN_W) // 2
        y = (self.winfo_screenheight() - height) // 2
        self.geometry(f"{self.WIN_W}x{height}+{x}+{y}")


if __name__ == "__main__":
    app = TemperatureConverter()
    app.mainloop()

File: lab5/lab5_1/Java/lab5_1.java
import java.awt.*;
import java.awt.event.*;
import javax.swing.*;
import javax.swing.border.*;

public class lab5_1 extends JFrame {

    // ── Minimal Modern Palette ────────────────────────────────────────────────
    private static final Color BG         = new Color(0x10, 0x10, 0x14);
    private static final Color PANEL      = new Color(0x17, 0x18, 0x1f);
    private static final Color PANEL_2    = new Color(0x1d, 0x1f, 0x27);
    private static final Color BORDER     = new Color(0x2a, 0x2d, 0x37);
    private static final Color ACCENT     = new Color(0x7c, 0x8c, 0xff);
    private static final Color ACCENT_HOV = new Color(0x93, 0xa0, 0xff);
    private static final Color TEXT       = new Color(0xf3, 0xf4, 0xf6);
    private static final Color TEXT_DIM   = new Color(0x9a, 0xa0, 0xad);
    private static final Color SUCCESS    = new Color(0x4a, 0xde, 0x80);
    private static final Color WARN       = new Color(0xfa, 0xcc, 0x15);
    private static final Color ERROR      = new Color(0xf8, 0x71, 0x71);
    private static final Color ENTRY_BG   = new Color(0x11, 0x13, 0x19);

    private static final int WIN_W = 420;
    private static final int WIN_H_NORMAL = 430;
    private static final int WIN_H_RESULT = 560;

    private String mode = "A";

    private JLabel inputLabel;
    private JTextField entryField;
    private JLabel errLabel;

    private JPanel resultFrame;
    private JLabel statusLabel;
    private JLabel resultValueLabel;
    private JLabel resultUnitLabel;
    private JLabel detailLabel;
    private JButton resetBtn;

    private SegmentedToggle toggle;

    public lab5_1() {
        setTitle("Temperature Converter");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setResizable(false);

        buildUI();
        setWindowHeight(WIN_H_NORMAL);
        centerWindow();
        setVisible(true);
    }

    private void buildUI() {
        JPanel outer = new JPanel(new BorderLayout());
        outer.setBackground(BG);
        outer.setBorder(new EmptyBorder(24, 24, 24, 24));
        setContentPane(outer);

        JPanel card = new JPanel();
        card.setLayout(new BoxLayout(card, BoxLayout.Y_AXIS));
        card.setBackground(PANEL);
        card.setBorder(new CompoundBorder(
                new LineBorder(BORDER, 1),
                new EmptyBorder(0, 0, 0, 0)
        ));
        outer.add(card, BorderLayout.CENTER);

        JLabel title = new JLabel("Temperature Converter");
        title.setForeground(TEXT);
        title.setFont(new Font("Segoe UI", Font.BOLD, 20));
        title.setBorder(new EmptyBorder(22, 24, 4, 24));
        title.setAlignmentX(Component.LEFT_ALIGNMENT);
        card.add(title);

        JLabel subtitle = new JLabel("Convert values instantly with a clean minimal interface.");
        subtitle.setForeground(TEXT_DIM);
        subtitle.setFont(new Font("Segoe UI", Font.PLAIN, 9));
        subtitle.setBorder(new EmptyBorder(0, 24, 18, 24));
        subtitle.setAlignmentX(Component.LEFT_ALIGNMENT);
        card.add(subtitle);

        JPanel divider = new JPanel();
        divider.setMaximumSize(new Dimension(Integer.MAX_VALUE, 1));
        divider.setPreferredSize(new Dimension(0, 1));
        divider.setBackground(BORDER);
        card.add(wrap(divider, PANEL, 0, 24, 18, 24));

        JLabel conversionLabel = new JLabel("Conversion");
        conversionLabel.setForeground(TEXT_DIM);
        conversionLabel.setFont(new Font("Segoe UI", Font.PLAIN, 9));
        conversionLabel.setBorder(new EmptyBorder(0, 24, 0, 24));
        conversionLabel.setAlignmentX(Component.LEFT_ALIGNMENT);
        card.add(conversionLabel);

        toggle = new SegmentedToggle(this::onModeChanged);
        toggle.setAlignmentX(Component.LEFT_ALIGNMENT);
        card.add(wrap(toggle, PANEL, 8, 24, 18, 24));

        inputLabel = new JLabel("Temperature (°C)");
        inputLabel.setForeground(TEXT_DIM);
        inputLabel.setFont(new Font("Segoe UI", Font.PLAIN, 9));
        inputLabel.setBorder(new EmptyBorder(0, 24, 0, 24));
        inputLabel.setAlignmentX(Component.LEFT_ALIGNMENT);
        card.add(inputLabel);

        JPanel entryShell = new JPanel(new BorderLayout());
        entryShell.setBackground(ENTRY_BG);
        entryShell.setBorder(new LineBorder(BORDER, 1));
        entryShell.setMaximumSize(new Dimension(Integer.MAX_VALUE, 50));

        entryField = new JTextField();
        entryField.setBackground(ENTRY_BG);
        entryField.setForeground(TEXT);
        entryField.setCaretColor(TEXT);
        entryField.setBorder(new EmptyBorder(14, 14, 14, 14));
        entryField.setFont(new Font("Segoe UI", Font.PLAIN, 16));
        entryField.addActionListener(e -> convert());

        entryField.addFocusListener(new FocusAdapter() {
            @Override
            public void focusGained(FocusEvent e) {
                entryShell.setBorder(new LineBorder(ACCENT, 1));
            }

            @Override
            public void focusLost(FocusEvent e) {
                entryShell.setBorder(new LineBorder(BORDER, 1));
            }
        });

        entryShell.add(entryField, BorderLayout.CENTER);
        card.add(wrap(entryShell, PANEL, 8, 24, 8, 24));

        errLabel = new JLabel(" ");
        errLabel.setForeground(ERROR);
        errLabel.setFont(new Font("Segoe UI", Font.PLAIN, 9));
        errLabel.setBorder(new EmptyBorder(0, 24, 12, 24));
        errLabel.setAlignmentX(Component.LEFT_ALIGNMENT);
        card.add(errLabel);

        JButton convertBtn = createButton("Convert", true, this::convert);
        card.add(wrap(convertBtn, PANEL, 0, 24, 18, 24));

        resultFrame = new JPanel();
        resultFrame.setLayout(new BoxLayout(resultFrame, BoxLayout.Y_AXIS));
        resultFrame.setBackground(PANEL_2);
        resultFrame.setBorder(new CompoundBorder(
                new LineBorder(BORDER, 1),
                new EmptyBorder(16, 18, 16, 18)
        ));
        resultFrame.setAlignmentX(Component.LEFT_ALIGNMENT);

        JPanel topRow = new JPanel(new BorderLayout());
        topRow.setOpaque(false);

        JLabel resultText = new JLabel("Result");
        resultText.setForeground(TEXT_DIM);
        resultText.setFont(new Font("Segoe UI", Font.PLAIN, 9));
        topRow.add(resultText, BorderLayout.WEST);

        statusLabel = new JLabel("");
        statusLabel.setForeground(WARN);
        statusLabel.setFont(new Font("Segoe UI", Font.BOLD, 9));
        topRow.add(statusLabel, BorderLayout.EAST);

        resultFrame.add(topRow);
        resultFrame.add(Box.createVerticalStrut(6));

        JPanel valueRow = new JPanel();
        valueRow.setLayout(new BoxLayout(valueRow, BoxLayout.X_AXIS));
        valueRow.setOpaque(false);

        resultValueLabel = new JLabel("");
        resultValueLabel.setForeground(TEXT);
        resultValueLabel.setFont(new Font("Segoe UI", Font.BOLD, 28));
        valueRow.add(resultValueLabel);

        valueRow.add(Box.createHorizontalStrut(8));

        resultUnitLabel = new JLabel("");
        resultUnitLabel.setForeground(ACCENT);
        resultUnitLabel.setFont(new Font("Segoe UI", Font.BOLD, 13));
        valueRow.add(resultUnitLabel);

        resultFrame.add(valueRow);
        resultFrame.add(Box.createVerticalStrut(6));

        detailLabel = new JLabel("");
        detailLabel.setForeground(TEXT_DIM);
        detailLabel.setFont(new Font("Segoe UI", Font.PLAIN, 9));
        resultFrame.add(detailLabel);

        resultFrame.setVisible(false);
        card.add(wrap(resultFrame, PANEL, 0, 24, 12, 24));

        resetBtn = createButton("Reset", false, this::reset);
        resetBtn.setVisible(false);
        card.add(wrap(resetBtn, PANEL, 0, 24, 22, 24));
    }

    private JPanel wrap(JComponent component, Color bg, int top, int left, int bottom, int right) {
        JPanel wrapper = new JPanel(new BorderLayout());
        wrapper.setBackground(bg);
        wrapper.setBorder(new EmptyBorder(top, left, bottom, right));
        wrapper.setAlignmentX(Component.LEFT_ALIGNMENT);
        wrapper.add(component, BorderLayout.CENTER);
        return wrapper;
    }

    private JButton createButton(String text, boolean primary, Runnable action) {
        JButton button = new JButton(text);
        button.setFocusPainted(false);
        button.setFont(new Font("Segoe UI", Font.BOLD, 10));
        button.setCursor(new Cursor(Cursor.HAND_CURSOR));
        button.setPreferredSize(new Dimension(0, 44));

        if (primary) {
            button.setBackground(ACCENT);
            button.setForeground(Color.WHITE);
            button.setBorder(new LineBorder(ACCENT, 1));
        } else {
            button.setBackground(PANEL_2);
            button.setForeground(TEXT);
            button.setBorder(new LineBorder(BORDER, 1));
        }

        button.setOpaque(true);
        button.setContentAreaFilled(true);

        button.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseEntered(MouseEvent e) {
                button.setBackground(primary ? ACCENT_HOV : new Color(0x25, 0x28, 0x33));
            }

            @Override
            public void mouseExited(MouseEvent e) {
                button.setBackground(primary ? ACCENT : PANEL_2);
            }
        });

        button.addActionListener(e -> action.run());
        return button;
    }

    private void onModeChanged(String newMode) {
        mode = newMode;

        if (errLabel != null) {
            errLabel.setText(" ");
        }

        hideResult();

        if (inputLabel != null) {
            if ("A".equals(mode)) {
                inputLabel.setText("Temperature (°C)");
            } else {
                inputLabel.setText("Temperature (°F)");
            }
        }
    }

    private void convert() {
        String raw = entryField.getText().trim();

        if (raw.isEmpty()) {
            errLabel.setText("Enter a temperature value.");
            hideResult();
            return;
        }

        double value;
        try {
            value = Double.parseDouble(raw);
        } catch (NumberFormatException ex) {
            errLabel.setText("Please enter a valid number.");
            hideResult();
            return;
        }

        errLabel.setText(" ");

        double converted;
        String outUnit;
        String detail;

        if ("A".equals(mode)) {
            converted = (value * 9.0 / 5.0) + 32.0;
            outUnit = "°F";
            detail = String.format("%.2f °C converted to Fahrenheit", value);
        } else {
            converted = (value - 32.0) * 5.0 / 9.0;
            outUnit = "°C";
            detail = String.format("%.2f °F converted to Celsius", value);
        }

        if (converted > 100) {
            statusLabel.setText("Too High");
            statusLabel.setForeground(WARN);
        } else {
            statusLabel.setText("Normal");
            statusLabel.setForeground(SUCCESS);
        }

        resultValueLabel.setText(String.format("%.2f", converted));
        resultUnitLabel.setText(outUnit);
        detailLabel.setText(detail);

        showResult();
    }

    private void showResult() {
        resultFrame.setVisible(true);
        resetBtn.setVisible(true);
        setWindowHeight(WIN_H_RESULT);
        centerWindow();
        revalidate();
        repaint();
    }

    private void hideResult() {
        if (resultFrame != null) {
            resultFrame.setVisible(false);
        }
        if (resetBtn != null) {
            resetBtn.setVisible(false);
        }
        setWindowHeight(WIN_H_NORMAL);
        centerWindow();
        revalidate();
        repaint();
    }

    private void reset() {
        entryField.setText("");
        errLabel.setText(" ");
        resultValueLabel.setText("");
        resultUnitLabel.setText("");
        detailLabel.setText("");
        statusLabel.setText("");
        entryField.requestFocusInWindow();
        hideResult();
    }

    private void setWindowHeight(int height) {
        setSize(WIN_W, height);
    }

    private void centerWindow() {
        Dimension screen = Toolkit.getDefaultToolkit().getScreenSize();
        int x = (screen.width - getWidth()) / 2;
        int y = (screen.height - getHeight()) / 2;
        setLocation(x, y);
    }

    public static void main(String[] args) {
        SwingUtilities.invokeLater(lab5_1::new);
    }

    private static class SegmentedToggle extends JPanel {
        private final JLabel btnA;
        private final JLabel btnB;
        private final ToggleCallback callback;
        private String mode = "A";

        SegmentedToggle(ToggleCallback callback) {
            this.callback = callback;

            setLayout(new GridLayout(1, 2));
            setBackground(PANEL);
            setBorder(new LineBorder(BORDER, 1));

            btnA = createSegment("°C → °F");
            btnB = createSegment("°F → °C");

            add(btnA);
            add(btnB);

            setModeSilently("A");

            btnA.addMouseListener(new MouseAdapter() {
                @Override
                public void mouseClicked(MouseEvent e) {
                    setMode("A");
                }
            });

            btnB.addMouseListener(new MouseAdapter() {
                @Override
                public void mouseClicked(MouseEvent e) {
                    setMode("B");
                }
            });
        }

        private JLabel createSegment(String text) {
            JLabel label = new JLabel(text, SwingConstants.CENTER);
            label.setOpaque(true);
            label.setCursor(new Cursor(Cursor.HAND_CURSOR));
            label.setFont(new Font("Segoe UI", Font.BOLD, 10));
            label.setBorder(new EmptyBorder(10, 24, 10, 24));
            return label;
        }

        private void applyModeStyles() {
            if ("A".equals(mode)) {
                btnA.setBackground(ACCENT);
                btnA.setForeground(Color.WHITE);
                btnB.setBackground(PANEL);
                btnB.setForeground(TEXT_DIM);
            } else {
                btnA.setBackground(PANEL);
                btnA.setForeground(TEXT_DIM);
                btnB.setBackground(ACCENT);
                btnB.setForeground(Color.WHITE);
            }
        }

        private void setModeSilently(String newMode) {
            mode = newMode;
            applyModeStyles();
        }

        private void setMode(String newMode) {
            mode = newMode;
            applyModeStyles();

            if (callback != null) {
                callback.onToggle(mode);
            }
        }
    }

    @FunctionalInterface
    private interface ToggleCallback {
        void onToggle(String mode);
    }
}